File: 391.go
package leetcode

import "math"

func isRectangleCover(rectangles [][]int) bool {
	area := 0
	left, bottom := math.MaxInt64, math.MaxInt64
	right, top := math.MinInt64, math.MinInt64
	cnt := map[[2]int]int{}
	for _, r := range rectangles {
		x, y, a, b := r[0], r[1], r[2], r[3]
		area += (a - x) * (b - y)
		left = min(left, x)
		bottom = min(bottom, y)
		right = max(right, a)
		top = max(top, b)

		cnt[[2]int{x, y}] += 1
		cnt[[2]int{x, b}] += 1
		cnt[[2]int{a, y}] += 1
		cnt[[2]int{a, b}] += 1
	}
	if area != (right-left)*(top-bottom) {
		return false
	}

	if cnt[[2]int{left, bottom}] != 1 {
		return false
	} else {
		delete(cnt, [2]int{left, bottom})
	}

	if cnt[[2]int{left, top}] != 1 {
		return false
	} else {
		delete(cnt, [2]int{left, top})
	}

	if cnt[[2]int{right, bottom}] != 1 {
		return false
	} else {
		delete(cnt, [2]int{right, bottom})
	}

	if cnt[[2]int{right, top}] != 1 {
		return false
	} else {
		delete(cnt, [2]int{right, top})
	}
	for _, val := range cnt {
		if !(val == 2 || val == 4) {
			return false
		}
	}
	return true
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 391.py
from typing import List


class Solution:
    def isRectangleCover(self, rectangles: List[List[int]]) -> bool:
        total_area = 0
        left = bottom = float('inf')
        right = top = float('-inf')
        cnt = collections.defaultdict(int)

        for x, y, a, b in rectangles:
            total_area += (a - x) * (b - y)
            left = min(left, x)
            bottom = min(bottom, y)
            right = max(right, a)
            top = max(top, b)

            cnt[(x, y)] += 1
            cnt[(x, b)] += 1
            cnt[(a, y)] += 1
            cnt[(a, b)] += 1

        # print(total_area, left, right, bottom, top)
        if total_area != (right - left) * (top - bottom) or cnt[(left, bottom)] != 1 \
                or cnt[(left, top)] != 1 or cnt[(right, bottom)] != 1 or cnt[(right, top)] != 1:
            return False
        del cnt[(left, bottom)], cnt[(left, top)], cnt[(
            right, bottom)], cnt[(right, top)]
        return all(c == 2 or c == 4 for c in cnt.values())
